src/kubernetes_charm.py

```diff
@@ -80,11 +97,136 @@ def _upgrade(self) -> typing.Optional[kubernetes_upgrade.Upgrade]:
+                name=self.service_name,
+                namespace=self.model.name,
+            )
+        except lightkube.core.exceptions.ApiError as e:
+            if e.status.code == 404:
+                return None
+            raise
+
+        return service
+
+    @functools.cache
+    def _get_pod(self, unit_name: str) -> lightkube.resources.core_v1.Pod:
+        """Get the pod for the provided unit name."""
+        return self._lightkube_client.get(
+            res=lightkube.resources.core_v1.Pod,
+            name=unit_name.replace("/", "-"),
+            namespace=self.model.name,
+        )
+
+    @functools.cache
+    def _get_node(self, unit_name: str) -> lightkube.resources.core_v1.Node:
+        """Return the node for the provided unit name."""
+        node_name = self._get_pod(unit_name).spec.nodeName
+        return self._lightkube_client.get(
+            res=lightkube.resources.core_v1.Node,
+            name=node_name,
+            namespace=self.model.name,
+        )
+
+    def _apply_service(self, service_type: _ServiceType, service_exists: bool) -> None:
+        """Apply the service type provided."""
+        pod0 = self._get_pod(f"{self.app.name}/0")
+
+        service = lightkube.resources.core_v1.Service(
+            metadata=lightkube.models.meta_v1.ObjectMeta(
+                name=self.service_name,
+                namespace=self.model.name,
+                ownerReferences=pod0.metadata.ownerReferences,  # the stateful set
+                labels={"app.kubernetes.io/name": self.app.name},
+            ),
+            spec=lightkube.models.core_v1.ServiceSpec(
+                ports=[
+                    lightkube.models.core_v1.ServicePort(
+                        name="mysql-rw",
+                        port=self._READ_WRITE_PORT,
+                        targetPort=self._READ_WRITE_PORT,
+                    ),
+                    lightkube.models.core_v1.ServicePort(
+                        name="mysql-ro",
+                        port=self._READ_ONLY_PORT,
+                        targetPort=self._READ_ONLY_PORT,
+                    ),
+                ],
+                type=service_type.value,
+                selector={"app.kubernetes.io/name": self.app.name},
+            ),
+        )
+
+        # Delete and re-create until https://bugs.launchpad.net/juju/+bug/2084711 resolved
+        if service_exists:
+            logger.info(f"Deleting service {service_type=}")
+            self._lightkube_client.delete(
+                res=lightkube.resources.core_v1.Service,
+                name=self.service_name,
+                namespace=self.model.name,
+            )
+            logger.info(f"Deleted service {service_type=}")
+
+        logger.info(f"Applying service {service_type=}")
+        self._lightkube_client.apply(service, field_manager=self.app.name)
+        logger.info(f"Applied service {service_type=}")
+
+    def _reconcile_service(self) -> None:
```

Comment:
Any reason to not merge `_apply_service` into `_reconcile_service`?

Reply:
no reason, merged in [f938351](https://github.com/canonical/mysql-router-k8s-operator/pull/328/commits/f9383516983d97e4d0cd34eb0c020ead40b08539)